File: Leetcode-solutions/Binary-search/ValidPerfectSquare.java
public class ValidPerfectSquare {
    public static void main(String[] args) {
        int num = 1;
        System.out.println(isPerfectSquare(num));
    }

    static boolean isPerfectSquare(int num) {
        long s = 1;
        long e = (num/2) +1;

        while(s <= e){
            long m = s + (e-s)/2;

            if(m*m == num){
                return true;
            }
            if(m *m < num){
                s = m + 1;
            } else{
                e = m-1;
            }
        }
        return false;
    }
}

File: Leetcode-solutions/Binary-search/ValidPerfectSquare.cpp
// Problem link : https://leetcode.com/problems/valid-perfect-square/
// Time Complexity : O(log(n))

class Solution {
public:
    
    int findSqrt(int num)
    {
        int low = 1;
        int high = num;
        
        int ans;
        
        while(low <= high){
            long long int mid = low + (high - low)/2;
            if(mid * mid <= num){
                ans = mid;
                low = mid + 1;
            }
            else{ // ( mid * mid ) is greater than num
                high = mid - 1;
            }
        }
        
        return ans;
    }
    
    bool isPerfectSquare(int num) {
        
        int sqrtN = findSqrt(num);
        
        if((sqrtN * sqrtN) == num) return true;
        
        return false;
    }
};